### PR TITLE
elasticsearch: Replace 'config' field with 'minimumMasters'

### PR DESCRIPTION
**Pull request by munnerz**
**What this PR does / why we need it**:

This PR alters the API surface for ElasticsearchCluster. It replaces the config field on node pools with a cluster wide minimum masters field.

The official Elastic recommendation is to modify the majority of cluster settings through the cluster API, and without a specific request to have a method to specify *static* config, I think we're best to remove this field for now.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #
 
Related to #194 

**Release note**:
```release-note
Remove 'config' field from node pools and replace with cluster-wide 'minimumMasters' field
```

### CODE REVIEW

File: pkg/apis/navigator/validation/elasticsearch_test.go [hunk 1]
```diff
@@ -312,52 +332,58 @@ func TestValidateElasticsearchPersistence(t *testing.T) {
...
+			Pilot:          validSpecPilotImage,
+			MinimumMasters: 2,
+		},
+		"minimum masters set to low": {
+			Plugins:        validSpecPluginsList,
+			NodePools:      []navigator.ElasticsearchClusterNodePool{newValidNodePool("test", 3, navigator.ElasticsearchRoleMaster)},
+			Pilot:          validSpecPilotImage,
+			Image:          validSpecESImage,
+			MinimumMasters: 1,
+		},
```
Comment (wallrj) [inline pkg/apis/navigator/validation/elasticsearch_test.go]: I wonder if there's ever a case where you want to run with fewer than quorum masters.
I mean if we're allowing a cluster with a single master why not allow a cluster, maybe for testing, with 3 masters but where only one is required to be running?

File: pkg/controllers/elasticsearch/configmap/resources.go [hunk 2]
```diff
@@ -16,6 +34,49 @@ func esConfigConfigMap(c *v1alpha1.ElasticsearchCluster, np *v1alpha1.Elasticsea
...
+	minimumMasters := c.Spec.MinimumMasters
+	if minimumMasters == 0 {
+		// auto-manage minimum master count
+		totalMasters := countMasterReplicas(c.Spec.NodePools)
+		minimumMasters = calculateQuorom(totalMasters)
+	}
+	return fmt.Sprintf(configTemplate, minimumMasters)
+}
+
+// TODO: move these functions into a shared package
```
Comment (wallrj) [inline pkg/controllers/elasticsearch/configmap/resources.go]: Yeah, I think we should do that before merging this.

File: pkg/apis/navigator/v1alpha1/types.go [hunk 3]
```diff
@@ -122,6 +122,12 @@ type ElasticsearchClusterSpec struct {
 	// Sysctl can be used to specify a list of sysctl values to set on start-up
 	// This can be used to set for example the vm.max_map_count parameter.
 	Sysctl []string `json:"sysctl"`
+
+	// The minimum number of masters required to form a quorom in the cluster.
+	// If omitted, this will be set to a quorom of the master nodes in the
+	// cluster. If set, the value *must* be greater than or equal to a quorom
+	// of master nodes.
```
[the remark this answers is not in the document]
Comment (munnerz) [reply pkg/apis/navigator/v1alpha1/types.go]: No - the field `minimumMasters` must be greater than or equal to a quorum of master nodes. If it were less than, a split brain could occur.

> And if so, how can you have a quorum greater than the number of masters in the cluster?

Exactly - if a user were to specify a `minimumMasters` parameter that is greater than the total number of masters in the cluster, a leader would never be elected.

File: pkg/apis/navigator/validation/elasticsearch.go [hunk 4]
```diff
@@ -96,9 +98,53 @@ func ValidateElasticsearchClusterSpec(spec *navigator.ElasticsearchClusterSpec,
...
+	case numMasters == 0:
+		allErrs = append(allErrs, field.Invalid(npPath, numMasters, "must be at least one master node"))
+	case spec.MinimumMasters == 0:
+		// do nothing, navigator-controller will automatically calculate &
+		// manage the minimumMasters required for the cluster
+	case spec.MinimumMasters < quorom:
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("minimumMasters"), spec.MinimumMasters, fmt.Sprintf("must be a minimum of %d to avoid a split brain scenario", quorom)))
+	case spec.MinimumMasters > quorom:
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("minimumMasters"), spec.MinimumMasters, fmt.Sprintf("cannot be greater than the total number of master nodes")))
+	}
```
[the remark this answers is not in the document]
Comment (munnerz) [reply pkg/apis/navigator/validation/elasticsearch.go]: 👍

File: pkg/apis/navigator/validation/elasticsearch_test.go (hunk 1 above)
Comment (munnerz) [reply pkg/apis/navigator/validation/elasticsearch_test.go]: This should never be allowed with Elasticsearch - I've created clusters with 3 nodes and a minimum of 1 master, and immediately got a split brain (two nodes elected themselves leader)

File: pkg/controllers/elasticsearch/configmap/resources.go (hunk 2 above)
Comment (munnerz) [reply pkg/controllers/elasticsearch/configmap/resources.go]: I've put something together for this, but it won't actually reduce duplication between validation and this package as they work on different types.

Validation is performed on internal types, and due to lack of generics, there's no way to share these functions (except for CalculateQuorum).

Comment (munnerz) [reply pkg/controllers/elasticsearch/configmap/resources.go]: I've pushed up my changes which include this. For now, I've kept the two functions that work on API types in validation, and also created `pkg/util/api` to house the v1alpha1 versions of these functions (which the `configmap` package now depends upon).

I did consider converting the types using the generated conversion functions, but this seems excessive for what is effectively a couple of for loops.

**Pull request merged**